File: methods/python/gp/gaussian_process.py
"""
Gaussian Process Methods for Derivative Estimation

This module implements GP-based derivative estimation methods using:
1. RBF (Squared Exponential) kernel with closed-form derivatives
2. Matérn kernel (nu=0.5, 1.5, 2.5) with analytical derivative formulas

Both methods use the posterior mean of the GP to estimate derivatives at arbitrary orders.
"""

from typing import Dict
import numpy as np
from numpy.polynomial import hermite_e as herme

from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.gaussian_process.kernels import RBF, WhiteKernel, ConstantKernel, Matern

# Import base class from common utilities
import sys
from pathlib import Path
import os
sys.path.insert(0, str(Path(__file__).parent))
from common import MethodEvaluator

# Optional robust optimizer: Powell → L-BFGS-B
try:
	from scipy.optimize import minimize  # type: ignore
	def _powell_then_lbfgsb(obj_func, initial_theta, bounds):
		f = lambda th: obj_func(th)[0]
		res1 = minimize(f, initial_theta, method="Powell", bounds=bounds,
						options={"maxiter": 2000, "xtol": 1e-4, "ftol": 1e-4})
		f2 = lambda th: obj_func(th)
		res2 = minimize(lambda th: f2(th)[0], res1.x,
						jac=lambda th: f2(th)[1],
						method="L-BFGS-B", bounds=bounds,
						options={"maxiter": 1000, "ftol": 1e-9})
		# scikit-learn expects (theta_opt, func_min)
		val, _grad = f2(res2.x)
		return res2.x, float(val)
	_OPTIMIZER_AVAILABLE = True
except Exception:
	_powell_then_lbfgsb = None
	_OPTIMIZER_AVAILABLE = False


def _heuristic_inits(x: np.ndarray, y: np.ndarray):
	# Use median pairwise distance for length scale (robust to outliers)
	dists = np.abs(np.subtract.outer(x, x))
	nonzero_dists = dists[dists > 0]
	if len(nonzero_dists) > 0:
		ell0 = float(np.median(nonzero_dists)) / np.sqrt(2)
	else:
		ell0 = 1.0

	# Use variance for amplitude
	amp0 = float(np.var(y)) if float(np.var(y)) > 0 else 1.0
	return ell0, amp0


def _build_gpr(kernel, use_powell: bool, restarts: int, normalize_y: bool = True):
	optimizer = _powell_then_lbfgsb if (use_powell and _OPTIMIZER_AVAILABLE) else "fmin_l_bfgs_b"
	return GaussianProcessRegressor(
		kernel=kernel,
		n_restarts_optimizer=restarts,
		alpha=1e-10,
		normalize_y=normalize_y,
		optimizer=optimizer,
	)


class GPMethods(MethodEvaluator):
	"""
	Gaussian Process methods for derivative estimation.

	Inherits from base MethodEvaluator and implements GP-specific methods.
	"""

	def evaluate_method(self, method_name: str) -> Dict:
		"""
		Dispatch to the appropriate GP method based on method name.

		Args:
			method_name: Name of method to evaluate. Supported:
				- "gp_rbf_mean", "GP_RBF_Python", "GP_RBF_Iso_Python": RBF kernel GP
				- "GP_Matern_Python", "GP_Matern_1.5_Python": Matérn(1.5) kernel GP
				- "GP_Matern_2.5_Python": Matérn(2.5) kernel GP

		Returns:
			Dictionary with predictions, failures, and metadata
		"""
		if method_name in ["gp_rbf_mean", "GP_RBF_Python", "GP_RBF_Iso_Python"]:
			return self._gp_rbf_mean_derivative()
		elif method_name in ["GP_Matern_Python", "GP_Matern_1.5_Python"]:
			return self._gp_matern(nu=1.5)
		elif method_name == "GP_Matern_2.5_Python":
			return self._gp_matern(nu=2.5)
		else:
			return {
				"predictions": {order: [np.nan] * len(self.x_eval) for order in self.orders},
				"failures": {"error": f"Unknown GP method: {method_name}"}
			}

	def _gp_rbf_mean_derivative(self) -> Dict:
		"""
		GP posterior mean derivatives using closed-form RBF kernel derivatives.

		Uses kernel: ConstantKernel * RBF + WhiteKernel
		Computes derivatives using Hermite polynomials for analytic kernel derivatives.

		Returns:
			Dictionary containing:
				predictions: Dict mapping derivative order to list of predictions
				failures: Dict of any errors encountered
				meta: Fitted hyperparameters (length_scale, amplitude)
		"""
		# Heuristic initial values and robust kernel with tighter bounds
		ell0, amp0 = _heuristic_inits(self.x_train, self.y_train)
		# Since normalize_y=True, amplitude should be ~1.0 (variance of normalized data)
		amp0 = 1.0
		# WhiteKernel upper bound set to 0.2 to handle up to ~20% noise on normalized data
		kernel = ConstantKernel(amp0, (1e-4, 1e4)) * RBF(length_scale=ell0, length_scale_bounds=(3e-3, 3e2)) \
				 + WhiteKernel(noise_level=1e-6, noise_level_bounds=(1e-10, 0.2))

		# Allow enabling Powell→LBFGSB via env var
		use_powell = str(os.environ.get("GP_OPTIMIZER", "")).lower() in ("powell", "powell_lbfgsb", "powell->lbfgsb")
		restarts = int(os.environ.get("GP_RESTARTS", "20")) if str(os.environ.get("GP_RESTARTS", "20")).isdigit() else 20

		gp = _build_gpr(kernel, use_powell, restarts)

		X = self.x_train.reshape(-1, 1)
		y = self.y_train
		gp.fit(X, y)

		# Extract fitted parameters (robust to structure Constant*RBF + White)
		fitted = gp.kernel_
		# Defaults
		amp = 1.0
		ell = 1.0
		try:
			# Expect Sum(Product(Constant, RBF), White)
			if hasattr(fitted, 'k1') and hasattr(fitted, 'k2'):
				k1 = fitted.k1
				# k1 should be Product(Constant, RBF)
				if hasattr(k1, 'k1') and hasattr(k1, 'k2') and isinstance(k1.k2, RBF):
					amp = float(getattr(k1.k1, 'constant_value', 1.0))
					ell = float(k1.k2.length_scale)
				elif isinstance(k1, RBF):
					ell = float(k1.length_scale)
			elif isinstance(fitted, RBF):
				ell = float(fitted.length_scale)
		except Exception:
			pass

		alpha = gp.alpha_.ravel()  # shape (n_train,)

		# Extract y normalization factors (if normalize_y=True)
		y_std = float(getattr(gp, "_y_train_std", 1.0))

		def herme_n(u: np.ndarray, n: int) -> np.ndarray:
			"""Evaluate probabilist's Hermite polynomial of degree n."""
			coeffs = np.zeros(n + 1)
			coeffs[-1] = 1.0
			return herme.hermeval(u, coeffs)

		def clamp_array(x: np.ndarray, lo: float, hi: float) -> np.ndarray:
			return np.clip(x, lo, hi)

		predictions = {}
		failures = {}

		Xtr = self.x_train
		for order in self.orders:
			try:
				if order == 0:
					preds = gp.predict(self.x_eval.reshape(-1, 1))
					predictions[order] = preds.astype(float).tolist()
					continue

				sign = -1.0 if (order % 2 == 1) else 1.0
				scale = (amp) * (ell ** (-order))
				# For each x*, compute k^(n)(x*, X) @ alpha
				out = []
				for xstar in self.x_eval:
					u = (xstar - Xtr) / ell  # shape (n_train,)
					# Clamp u to avoid overflow in Hermite and exp
					u_safe = clamp_array(u, -50.0, 50.0)
					base = np.exp(-0.5 * (u_safe ** 2))  # exp(-(x-x')^2/(2ℓ^2))
					hn = herme_n(u_safe, order)
					# Clamp scaling to prevent extreme magnitudes when ell is very small
					scale_raw = scale
					scale_safe = float(np.clip(scale_raw, -1e10, 1e10))
					k_n = (sign * scale_safe) * hn * base  # derivative wrt x*
					out.append(float(y_std * (k_n @ alpha)))
				predictions[order] = out
			except Exception as e:
				failures[order] = str(e)
				predictions[order] = [np.nan] * len(self.x_eval)

		return {"predictions": predictions, "failures": failures, "meta": {"length_scale": ell, "amplitude": amp}}

	def _gp_matern(self, nu: float) -> Dict:
		"""
		GP with Matérn kernel; posterior mean derivatives via closed-form kernel derivatives.

		OPTIMIZED VERSION: Uses analytical formulas for Matern kernel derivatives instead of
		nested autograd differentiation. This is 100-1000× faster for high-order derivatives.

		Args:
			nu: Smoothness parameter. Supported values: 0.5, 1.5, 2.5

		Returns:
			Dictionary containing:
				predictions: Dict mapping derivative order to list of predictions
				failures: Dict of any errors encountered
				meta: Fitted hyperparameters (nu, length_scale, amplitude)
		"""
		ell0, amp0 = _heuristic_inits(self.x_train, self.y_train)
		# Since normalize_y=True, amplitude should be ~1.0 (variance of normalized data)
		amp0 = 1.0
		# WhiteKernel upper bound set to 0.2 to handle up to ~20% noise on normalized data
		kernel = ConstantKernel(amp0, (1e-4, 1e4)) * Matern(length_scale=ell0, length_scale_bounds=(3e-3, 3e2), nu=nu) \
				 + WhiteKernel(noise_level=1e-6, noise_level_bounds=(1e-10, 0.2))

		use_powell = str(os.environ.get("GP_OPTIMIZER", "")).lower() in ("powell", "powell_lbfgsb", "powell->lbfgsb")
		restarts = int(os.environ.get("GP_RESTARTS", "20")) if str(os.environ.get("GP_RESTARTS", "20")).isdigit() else 20

		gp = _build_gpr(kernel, use_powell, restarts)
		X = self.x_train.reshape(-1, 1)
		y = self.y_train
		gp.fit(X, y)

		# Extract fitted params
		fitted = gp.kernel_
		amp = 1.0
		ell = 1.0
		noise = 1e-8
		try:
			if hasattr(fitted, 'k1') and hasattr(fitted, 'k2'):
				prod = fitted.k1
				if hasattr(prod, 'k1') and hasattr(prod, 'k2') and isinstance(prod.k2, Matern):
					amp = float(getattr(prod.k1, 'constant_value', 1.0))
					ell = float(prod.k2.length_scale)
					noise = float(getattr(fitted.k2, 'noise_level', 1e-8))
		except Exception:
			pass

		# Prepare alpha
		alpha = gp.alpha_.ravel()
		Xtr = self.x_train

		# Extract y normalization factors (if normalize_y=True)
		y_std = float(getattr(gp, "_y_train_std", 1.0))

		# Closed-form Matern kernel derivatives
		def matern_kernel_derivative(x: float, xprime: float, ell_local: float, nu_local: float, order: int) -> float:
			"""
			Compute order-th derivative of Matern kernel wrt x.

			Args:
				x: Evaluation point
				xprime: Training point
				ell_local: Length scale
				nu_local: Smoothness parameter
				order: Derivative order

			Returns:
				Kernel derivative value
			"""
			r = abs(x - xprime)
			r_safe = max(r, 1e-12)
			sign = 1.0 if x >= xprime else -1.0

			if abs(nu_local - 0.5) < 1e-8:
				# Matern-1/2: k^(n)(r) = (-1/ℓ)^n · exp(-r/ℓ)
				result = ((-1.0 / ell_local) ** order) * np.exp(-r_safe / ell_local)
				if order % 2 == 1:
					result *= sign

			elif abs(nu_local - 1.5) < 1e-8:
				# Matern-3/2: closed-form derivatives
				c = np.sqrt(3.0) / ell_local
				cr = c * r_safe
				exp_term = np.exp(-cr)

				if order == 0:
					result = (1.0 + cr) * exp_term
				elif order == 1:
					result = sign * (-c * c * r_safe) * exp_term
				elif order == 2:
					result = (c ** 2) * (cr - 1.0) * exp_term
				elif order == 3:
					result = sign * (c ** 3) * (3.0 - cr) * exp_term
				elif order == 4:
					result = (c ** 4) * (cr - 3.0) * exp_term
				elif order == 5:
					result = sign * (c ** 5) * (5.0 - cr) * exp_term
				elif order == 6:
					result = (c ** 6) * (cr - 5.0) * exp_term
				elif order == 7:
					result = sign * (c ** 7) * (7.0 - cr) * exp_term
				else:
					result = 0.0  # Unsupported order

			elif abs(nu_local - 2.5) < 1e-8:
				# Matern-5/2: closed-form derivatives
				c = np.sqrt(5.0) / ell_local
				cr = c * r_safe
				cr2 = cr * cr
				exp_term = np.exp(-cr)

				if order == 0:
					result = (1.0 + cr + cr2 / 3.0) * exp_term
				elif order == 1:
					result = sign * (c / 3.0) * cr * (cr - 3.0) * exp_term
				elif order == 2:
					result = (c ** 2 / 3.0) * (cr2 - 6.0 * cr + 3.0) * exp_term
				elif order == 3:
					result = sign * (c ** 3 / 3.0) * (cr2 - 9.0 * cr + 15.0) * exp_term
				elif order == 4:
					result = (c ** 4 / 3.0) * (cr2 - 12.0 * cr + 15.0) * exp_term
				elif order == 5:
					result = sign * (c ** 5 / 3.0) * (cr2 - 15.0 * cr + 45.0) * exp_term
				elif order == 6:
					result = (c ** 6 / 3.0) * (cr2 - 18.0 * cr + 45.0) * exp_term
				elif order == 7:
					result = sign * (c ** 7 / 3.0) * (cr2 - 21.0 * cr + 105.0) * exp_term
				else:
					result = 0.0  # Unsupported order
			else:
				# Fallback to RBF-like (not true Matern for general nu)
				if order == 0:
					result = np.exp(-0.5 * (r_safe / ell_local) ** 2)
				else:
					result = 0.0

			return result

		predictions = {}
		failures = {}
		for order in self.orders:
			try:
				if order == 0:
					mu = gp.predict(self.x_eval.reshape(-1, 1))
					predictions[order] = [float(v) for v in mu]
				else:
					# Use closed-form kernel derivatives (FAST!)
					vals = []
					for xstar in self.x_eval:
						deriv_sum = 0.0
						for xj, aj in zip(Xtr, alpha):
							k_deriv = matern_kernel_derivative(xstar, xj, ell, nu, order)
							deriv_sum += amp * k_deriv * aj
						vals.append(float(y_std * deriv_sum))
					predictions[order] = vals
			except Exception as e:
				failures[order] = str(e)
				predictions[order] = [np.nan] * len(self.x_eval)

		return {"predictions": predictions, "failures": failures, "meta": {"nu": nu, "length_scale": ell, "amplitude": amp}}

File: methods/julia/gp/gaussian_process.jl
"""
Gaussian Process Methods for Derivative Estimation

Implements 5 GP-based methods:
1. GP-Julia-SE: Analytic SE (RBF) kernel with closed-form derivatives
2. GP-Julia-AD: AD-based GP with generic kernels
3. GP-Julia-Matern-0.5: Matérn-1/2 kernel (exponential)
4. GP-Julia-Matern-1.5: Matérn-3/2 kernel (once differentiable)
5. GP-Julia-Matern-2.5: Matérn-5/2 kernel (twice differentiable)
"""

# Load common utilities
include("../common.jl")

using GaussianProcesses
using Optim
using LineSearches
using Dates  # For timestamp logging
# Note: Suppressor removed - @suppress macro causes threading issues (file descriptor conflicts)

# ============================================================================
# GP-Julia-SE: Analytic SE (RBF) GP with closed-form derivatives
# ============================================================================

"""
	fit_gp_se_analytic(x, y)

Fit a 1D SE GP by MLE and return a callable that yields n-th derivatives using
closed-form kernel derivatives (probabilists' Hermite polynomials).
"""
function fit_gp_se_analytic(x::Vector{Float64}, y::Vector{Float64})
	n = length(x)
	@assert n == length(y)

	# Center y to handle mean; derivatives of constant are zero
	y_mean = mean(y)
	yc = y .- y_mean

	# Normalize x for better conditioning
	x_mean = mean(x)
	x_std = std(x)
	x_std = x_std == 0.0 ? 1.0 : x_std
	x_scaled = (x .- x_mean) ./ x_std

	# Precompute pairwise squared distances matrix (for efficiency)
	function sqdist_matrix(xv::Vector{Float64})
		X = repeat(xv, 1, length(xv))
		return (X .- X') .^ 2
	end
	D2 = sqdist_matrix(x_scaled)

	# Robust Cholesky with escalating jitter (used both in NLL and final)
	function cholesky_with_jitter!(K::AbstractMatrix{T}) where T <: Real
		d = diag(K)
		# Use primal values to avoid ordering on Dual numbers
		dpr = map(x -> ForwardDiff.value(x), d)
		dmed = median(dpr)
		base = max(T(1e-12), T(1e-8) * T(dmed))
		jitter = base
		for _ in 1:12
			try
				return cholesky(Symmetric(K))
			catch
				@inbounds for i in 1:size(K, 1)
					K[i, i] += jitter
				end
				jitter *= T(10)
			end
		end
		return cholesky(Symmetric(K); check = false)
	end

	# Negative log marginal likelihood (with log-params) using robust PD handling
	function nll(p)
		ℓ = exp(p[1])
		σf = exp(p[2])
		σn = max(exp(p[3]), 1e-6)
		K = (σf^2) .* exp.(-0.5 .* (D2 ./ (ℓ^2)))
		@inbounds for i in 1:n
			K[i, i] += σn^2
		end
		# tiny base jitter before attempting factorization
		@inbounds for i in 1:n
			K[i, i] += 1e-12
		end
		F = cholesky_with_jitter!(K)
		α = F \ yc
		return 0.5 * dot(yc, α) + sum(log, diag(F.U)) + 0.5 * n * log(2π)
	end

	# Initial params from heuristics
	p0 = [log(std(x) / 8), log(std(yc) + eps()), log(std(yc) / 100 + 1e-6)]
	# Use finite-difference gradients to avoid Dual types inside Cholesky
	result = Optim.optimize(nll, p0, LBFGS(); autodiff = :finite)
	p̂ = Optim.minimizer(result)
	ℓ̂ = exp(p̂[1])
	σf̂ = exp(p̂[2])
	σn̂ = exp(p̂[3])

	# Precompute training structures
	K̂ = (σf̂^2) .* exp.(-0.5 .* (D2 ./ (ℓ̂^2)))
	@inbounds for i in 1:n
		K̂[i, i] += σn̂^2
	end
	F̂ = cholesky_with_jitter!(K̂)
	α̂ = F̂ \ yc

	# Probabilists' Hermite polynomials He_n(u)
	# FIX: Clamping u to prevent overflow in recursion
	function hermite_prob(n::Int, u::AbstractVector{T}) where {T}
		# Clamp input to prevent overflow (exp(-0.5*u^2) becomes negligible beyond |u|=50)
		u_safe = clamp.(u, T(-50), T(50))

		if n == 0
			return ones(T, length(u_safe))
		elseif n == 1
			return copy(u_safe)
		else
			He_nm1 = ones(T, length(u_safe))
			He_n0 = copy(u_safe)
			for k in 1:(n-1)
				He_np1 = u_safe .* He_n0 .- k .* He_nm1
				He_nm1, He_n0 = He_n0, He_np1
			end
			return He_n0
		end
	end

	# Vectorized evaluator for n-th derivative of posterior mean at x*
	function eval_nth_deriv(xstar::Float64, n::Int)
		if n == 0
			# Standard predictive mean: k_*^T α̂ + y_mean
			xstar_scaled = (xstar - x_mean) / x_std
			u = (xstar_scaled .- x_scaled) ./ ℓ̂
			kstar = (σf̂^2) .* exp.(-0.5 .* (u .^ 2))
			return dot(kstar, α̂) + y_mean
		else
			xstar_scaled = (xstar - x_mean) / x_std
			u = (xstar_scaled .- x_scaled) ./ ℓ̂

			# Check for numerical instability before computing
			if maximum(abs.(u)) > 10.0
				@warn "GP-Julia-SE: Large u values (max=$(maximum(abs.(u)))), order $n may be unstable"
			end

			base = exp.(-0.5 .* (u .^ 2))
			He = hermite_prob(n, u)
			sign = (n % 2 == 1) ? -1.0 : 1.0
			# Clamp scaling factor to prevent overflow
			scale_raw = (σf̂^2) * (ℓ̂ ^ (-n))
			scale = clamp(scale_raw, -1e10, 1e10)

			k_n = (sign .* scale) .* He .* base
			result = dot(k_n, α̂)

			# Return NaN if result is non-finite
			return isfinite(result) ? result : NaN
		end
	end

	return (z::Float64, n::Int = 0) -> eval_nth_deriv(z, n)
end


# ============================================================================
# GP-TaylorAD-Julia: AD-based GP with generic kernels
# ============================================================================

"""
	fit_gp(x, y; kernel=:SE, optimize=true)

Fit Gaussian Process with specified kernel.
"""
function fit_gp(x, y; kernel = :SE, optimize = true)
	# Scale data for numerical stability
	x_mean, x_std = mean(x), std(x)
	y_mean, y_std = mean(y), std(y)
	x_scaled = (x .- x_mean) ./ x_std
	y_scaled = (y .- y_mean) ./ max(y_std, 1e-8)

	# Optional tiny jitter to stabilize factorization
	y_jittered = y_scaled .+ 1e-8 .* randn(length(y_scaled))

	# Choose kernel (SE/RBF)
	kern = SEIso(0.0, 0.0)

	# Create GP with a modest noise floor
	gp = GP(x_scaled, y_jittered, MeanZero(), kern, log(1e-2))

	# Optimize hyperparameters (robust to failures)
	if optimize
		try
			optimize!(gp; method = LBFGS(linesearch = LineSearches.BackTracking()))
		catch e
			@warn "GP optimization failed, proceeding with current hyperparameters" exception=(e,)
		end
	end

	# AD-friendly predictor: use predict_f, not predict_y
	gp_func = z -> begin
		z_scaled = (z - x_mean) / x_std
		pred, _ = predict_f(gp, [z_scaled])
		return pred[1] * y_std + y_mean
	end

	return gp_func
end

"""
	fit_gp_ad(x, y; rng=Random.GLOBAL_RNG)

Fit GP using TaylorDiff-based automatic differentiation (simpler, more robust).
Based on ODEParameterEstimation approach - normalize, optimize, then AD through predictions.

Args:
	x: Input locations
	y: Observations
	rng: Random number generator for reproducible jitter (default: global RNG)
"""
function fit_gp_ad(x::Vector{Float64}, y::Vector{Float64}; rng = Random.GLOBAL_RNG)
	@assert length(x) == length(y) "Input arrays must have same length"

	# 1. Normalize y values
	y_mean = mean(y)
	y_std = std(y)
	y_normalized = (y .- y_mean) ./ max(y_std, 1e-8)  # Avoid division by very small numbers

	# Configure initial GP hyperparameters
	initial_lengthscale = log(std(x) / 8)
	initial_variance = 0.0
	initial_noise = -2.0

	# Add small amount of jitter to avoid numerical issues
	# IMPORTANT: Use provided RNG for reproducibility
	kernel = SEIso(initial_lengthscale, initial_variance)
	jitter = 1e-8
	y_jittered = y_normalized .+ jitter * randn(rng, length(y_normalized))

	# 2. Do GPR approximation on normalized data
	# Note: @suppress removed due to threading issues (file descriptor conflicts in parallel execution)
	local gp
	gp = GP(x, y_jittered, MeanZero(), kernel, initial_noise)
	GaussianProcesses.optimize!(gp; method = LBFGS(linesearch = LineSearches.BackTracking()))

	# Create a function that evaluates the GPR prediction and denormalizes the output
	function denormalized_gpr(x_eval)
		pred, _ = predict_f(gp, [x_eval])
		return y_std * pred[1] + y_mean
	end

	return denormalized_gpr
end


# ============================================================================
# GP-Julia-Matern: GP with Matérn kernels
# ============================================================================

"""
	fit_gp_matern(x, y; nu=1.5, rng=Random.GLOBAL_RNG)

Fit GP with Matérn kernel using TaylorDiff-based automatic differentiation.
Uses same approach as GP-Julia-AD but with Matérn kernel instead of SE.

Supported nu values: 0.5 (Mat12Iso), 1.5 (Mat32Iso), 2.5 (Mat52Iso)

Args:
	x: Input locations
	y: Observations
	nu: Matérn smoothness parameter
	rng: Random number generator for reproducible jitter (default: global RNG)
"""
function fit_gp_matern(x::Vector{Float64}, y::Vector{Float64}; nu::Float64 = 1.5, rng = Random.GLOBAL_RNG)
	@assert length(x) == length(y) "Input arrays must have same length"

	# 1. Normalize y values
	y_mean = mean(y)
	y_std = std(y)
	y_normalized = (y .- y_mean) ./ max(y_std, 1e-8)

	# Configure initial GP hyperparameters
	initial_lengthscale = log(std(x) / 8)
	initial_variance = 0.0
	initial_noise = -2.0

	# Select Matérn kernel based on nu
	kernel = if abs(nu - 0.5) < 1e-8
		Mat12Iso(initial_lengthscale, initial_variance)  # Matérn-1/2
	elseif abs(nu - 1.5) < 1e-8
		Mat32Iso(initial_lengthscale, initial_variance)  # Matérn-3/2
	elseif abs(nu - 2.5) < 1e-8
		Mat52Iso(initial_lengthscale, initial_variance)  # Matérn-5/2
	else
		error("Unsupported nu value: $nu (use 0.5, 1.5, or 2.5)")
	end

	# Add small amount of jitter to avoid numerical issues
	# IMPORTANT: Use provided RNG for reproducibility
	jitter = 1e-8
	y_jittered = y_normalized .+ jitter * randn(rng, length(y_normalized))

	# 2. Do GPR approximation on normalized data
	# Note: @suppress removed due to threading issues (file descriptor conflicts in parallel execution)
	local gp
	gp = GP(x, y_jittered, MeanZero(), kernel, initial_noise)
	GaussianProcesses.optimize!(gp; method = LBFGS(linesearch = LineSearches.BackTracking()))

	# Create a function that evaluates the GPR prediction and denormalizes the output
	function denormalized_gpr(x_eval)
		pred, _ = predict_f(gp, [x_eval])
		return y_std * pred[1] + y_mean
	end

	return denormalized_gpr
end


# ============================================================================
# Method Evaluators (Standard API)
# ============================================================================

"""
	evaluate_gp_se(x, y, x_eval, orders; params=Dict())

Evaluate GP-Julia-SE method: Analytic SE kernel with closed-form derivatives.
"""
function evaluate_gp_se(
	x::Vector{Float64},
	y::Vector{Float64},
	x_eval::Vector{Float64},
	orders::Vector{Int};
	params = Dict()
)
	t_start = time()
	predictions = Dict{Int, Vector{Float64}}()
	failures = Dict{Int, String}()

	try
		# Fit analytic SE GP
		gp_analytic = fit_gp_se_analytic(x, y)

		# Evaluate derivatives directly from analytic SE GP
		for order in orders
			try
				predictions[order] = [gp_analytic(xi, order) for xi in x_eval]
			catch e
				predictions[order] = fill(NaN, length(x_eval))
				failures[order] = string(e)
			end
		end

		timing = time() - t_start
		return MethodResult("GP-Julia-SE", "Gaussian Process", predictions, failures, timing, true)

	catch e
		@warn "GP-Julia-SE failed" exception=e
		timing = time() - t_start
		return MethodResult("GP-Julia-SE", "Error", Dict(), Dict(0 => string(e)), timing, false)
	end
end

"""
	evaluate_gp_ad(x, y, x_eval, orders; params=Dict())

Evaluate GP-Julia-AD method: AD-based GP with generic kernels.
"""
function evaluate_gp_ad(
	x::Vector{Float64},
	y::Vector{Float64},
	x_eval::Vector{Float64},
	orders::Vector{Int};
	params = Dict()
)
	t_start = time()
	predictions = Dict{Int, Vector{Float64}}()
	failures = Dict{Int, String}()

	# Setup verbose logging to gitignored file
	trial_id = get(params, :trial_id, "unknown")
	log_dir = joinpath(dirname(@__DIR__), "..", "..", "build", "logs", "gp")
	mkpath(log_dir)
	log_file = joinpath(log_dir, "$(trial_id).log")

	# Extract RNG from params for reproducibility
	rng = get(params, :rng, Random.GLOBAL_RNG)
	rng_seed = get(params, :rng_seed, "unknown")

	# Log trial start
	open(log_file, "w") do io
		println(io, "=" ^ 80)
		println(io, "GP-TaylorAD-Julia Detailed Log")
		println(io, "=" ^ 80)
		println(io, "Trial ID: $trial_id")
		println(io, "Timestamp: $(now())")
		println(io, "RNG Seed: $rng_seed")
		println(io)
		println(io, "Data Statistics:")
		println(io, "  Number of points: $(length(x))")
		println(io, "  x range: [$(minimum(x)), $(maximum(x))]")
		println(io, "  y range: [$(minimum(y)), $(maximum(y))]")
		println(io, "  y mean: $(mean(y))")
		println(io, "  y std: $(std(y))")
		println(io, "  Orders requested: $orders")
		println(io)
	end

	try
		# Fit AD-based GP
		fitted_func = fit_gp_ad(x, y; rng = rng)

		# Use TaylorDiff for automatic differentiation (all orders)
		for order in orders
			try
				predictions[order] = [nth_deriv_taylor(fitted_func, order, xi) for xi in x_eval]
			catch e
				predictions[order] = fill(NaN, length(x_eval))
				failures[order] = string(e)
			end
		end

		timing = time() - t_start

		# Log success
		open(log_file, "a") do io
			println(io, "Status: SUCCESS")
			println(io, "Total time: $(timing) seconds")
			println(io, "Orders computed: $(sort(collect(keys(predictions))))")
			if !isempty(failures)
				println(io, "Partial failures: $(sort(collect(keys(failures))))")
			end
			println(io, "=" ^ 80)
		end

		return MethodResult("GP-TaylorAD-Julia", "Gaussian Process", predictions, failures, timing, true)

	catch e
		# Enhanced error logging for debugging GP failures
		error_details = sprint(showerror, e, catch_backtrace())
		timing = time() - t_start

		# Log failure to file
		open(log_file, "a") do io
			println(io, "Status: FAILED")
			println(io, "Total time: $(timing) seconds")
			println(io)
			println(io, "Error Details:")
			println(io, error_details)
			println(io, "=" ^ 80)
		end

		# Also log to console
		@error "GP-TaylorAD-Julia FAILED" exception=(e, catch_backtrace()) data_stats=(
			n_points=length(x),
			x_range=(minimum(x), maximum(x)),
			y_range=(minimum(y), maximum(y)),
			y_std=std(y),
			y_mean=mean(y)
		)

		return MethodResult("GP-TaylorAD-Julia", "Error", Dict(), Dict(0 => error_details), timing, false)
	end
end

"""
	evaluate_gp_matern(x, y, x_eval, orders; nu=1.5, params=Dict())

Evaluate GP-Julia-Matern methods with specified nu parameter.
"""
function evaluate_gp_matern(
	x::Vector{Float64},
	y::Vector{Float64},
	x_eval::Vector{Float64},
	orders::Vector{Int};
	nu::Float64 = 1.5,
	params = Dict()
)
	t_start = time()
	predictions = Dict{Int, Vector{Float64}}()
	failures = Dict{Int, String}()

	# Determine method name based on nu
	method_name = "GP-Julia-Matern-$(nu)"

	try
		# Fit Matérn GP
		# Extract RNG from params for reproducibility
		rng = get(params, :rng, Random.GLOBAL_RNG)
		fitted_func = fit_gp_matern(x, y; nu = nu, rng = rng)

		# Use TaylorDiff for automatic differentiation (all orders)
		for order in orders
			try
				predictions[order] = [nth_deriv_taylor(fitted_func, order, xi) for xi in x_eval]
			catch e
				predictions[order] = fill(NaN, length(x_eval))
				failures[order] = string(e)
			end
		end

		timing = time() - t_start
		return MethodResult(method_name, "Gaussian Process", predictions, failures, timing, true)

	catch e
		@warn "$method_name failed" exception=e
		timing = time() - t_start
		return MethodResult(method_name, "Error", Dict(), Dict(0 => string(e)), timing, false)
	end
end

# Convenience wrappers for specific nu values
evaluate_gp_matern_05(x, y, x_eval, orders; params = Dict()) =
	evaluate_gp_matern(x, y, x_eval, orders; nu = 0.5, params = params)

evaluate_gp_matern_15(x, y, x_eval, orders; params = Dict()) =
	evaluate_gp_matern(x, y, x_eval, orders; nu = 1.5, params = params)

evaluate_gp_matern_25(x, y, x_eval, orders; params = Dict()) =
	evaluate_gp_matern(x, y, x_eval, orders; nu = 2.5, params = params)
